flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -795,20 +806,241 @@ private void restoreInstance(
+			}
+
+			@Override
+			public void close() throws Exception {
+
+				IOUtils.closeQuietly(defaultColumnFamilyHandle);
+
+				for (ColumnFamilyHandle columnFamilyHandle : columnFamilyHandles) {
+					IOUtils.closeQuietly(columnFamilyHandle);
+				}
+
+				IOUtils.closeQuietly(db);
+			}
+		}
+
+		private RestoredDBInstance restoreDBFromStateHandle(
+			IncrementalKeyedStateHandle restoreStateHandle,
+			Path temporaryRestoreInstancePath,
+			KeyGroupRange targetKeyGroupRange,
+			int keyGroupPrefixBytes,
+			boolean needClip) throws Exception {
+
+			transferAllStateDataToDirectory(restoreStateHandle, temporaryRestoreInstancePath);
+
 			// read meta data
 			List<RegisteredKeyedBackendStateMetaInfo.Snapshot<?, ?>> stateMetaInfoSnapshots =
-				readMetaData(localKeyedStateHandle.getMetaDataState());
+				readMetaData(restoreStateHandle.getMetaStateHandle());
 
 			List<ColumnFamilyDescriptor> columnFamilyDescriptors =
 				createAndRegisterColumnFamilyDescriptors(stateMetaInfoSnapshots);
 
-			restoreLocalStateIntoFullInstance(
-				localKeyedStateHandle,
+			List<ColumnFamilyHandle> columnFamilyHandles =
+				new ArrayList<>(stateMetaInfoSnapshots.size() + 1);
+
+			RocksDB restoreDb = stateBackend.openDB(
+				temporaryRestoreInstancePath.getPath(),
 				columnFamilyDescriptors,
-				stateMetaInfoSnapshots);
+				columnFamilyHandles);
+
+			if (needClip) {
+				RocksDBIncrementalCheckpointUtils.clipDBWithKeyGroupRange(
+					restoreDb,
+					columnFamilyHandles,
+					targetKeyGroupRange,
+					restoreStateHandle.getKeyGroupRange(),
+					keyGroupPrefixBytes);
+			}
+
+			return new RestoredDBInstance(restoreDb, columnFamilyHandles, columnFamilyDescriptors, stateMetaInfoSnapshots);
+		}
+
+		private ColumnFamilyHandle getOrRegisterColumnFamilyHandle(
+			ColumnFamilyDescriptor columnFamilyDescriptor,
+			ColumnFamilyHandle columnFamilyHandle,
+			RegisteredKeyedBackendStateMetaInfo.Snapshot<?, ?> stateMetaInfoSnapshot) throws RocksDBException {
+
+			Tuple2<ColumnFamilyHandle, RegisteredKeyedBackendStateMetaInfo<?, ?>> registeredStateMetaInfoEntry =
+				stateBackend.kvStateInformation.get(stateMetaInfoSnapshot.getName());
+
+			if (null == registeredStateMetaInfoEntry) {
+				RegisteredKeyedBackendStateMetaInfo<?, ?> stateMetaInfo =
+					new RegisteredKeyedBackendStateMetaInfo<>(
+						stateMetaInfoSnapshot.getStateType(),
+						stateMetaInfoSnapshot.getName(),
+						stateMetaInfoSnapshot.getNamespaceSerializer(),
+						stateMetaInfoSnapshot.getStateSerializer());
+
+				registeredStateMetaInfoEntry =
+					new Tuple2<>(
+						columnFamilyHandle != null ? columnFamilyHandle : stateBackend.db.createColumnFamily(columnFamilyDescriptor),
+						stateMetaInfo);
+
+				stateBackend.kvStateInformation.put(
+					stateMetaInfoSnapshot.getName(),
+					registeredStateMetaInfoEntry);
+			}
+
+			return registeredStateMetaInfoEntry.f0;
+		}
+
+		private void chooseTheBestStateHandleToInit(
```

Comment:
I think the name of this method is no longer accurate: it does not only chose the best handle, it already restores as db instance. Maybe we can we still break this up into two methods, so that each method only does one thing. I think it is not so nice if that creating the db is a side effect of a method that claims to only find something.

Reply:
Yes, I will split it into two methods.